### PR TITLE
Module cmd fix

### LINKED ISSUE #2924: install fails looking for `modulecmd`
A user reported this error.

It looks like [`load_module` in `build_enviroment.py`](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/build_environment.py#L129) will fail if `modulecmd` is not available.  

@alalazo: Is it supposed to be possible to get to this code if the system supports modules?  Seems like we should check for `modulecmd` up front.  I haven't dug into this yet.

```console
galaxy:~>spack -d -v mirror list                                                                      
==> Reading config file /local/cjn/.spack/linux/mirrors.yaml                                         
local_filesystem    file:///ROWAN/group/pts/spack_mirror/spack-mirror-2017-01-24

galaxy:~>spack -d -v install zlib
==> Reading config file /ROWAN/group/prod/LINUX_WORKSTATION/acs/spack/spack-0.10.0/etc/spack/defaults/packages.yaml
==> READ LOCK: /local/cjn/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Acquiring]
==> READ LOCK: /local/cjn/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Released]
==> Reading config file /local/cjn/.spack/linux/compilers.yaml
==> READ LOCK: /ROWAN/group/prod/LINUX_WORKSTATION/acs/spack/spack-0.10.0/opt/spack/.spack-db/prefix_lock[3332486004197984765:1] [Acquiring]
==> READ LOCK: /ROWAN/group/prod/LINUX_WORKSTATION/acs/spack/spack-0.10.0/opt/spack/.spack-db/prefix_lock[3332486004197984765:1] [Released]
==> Installing zlib
==> Error: AttributeError: 'NoneType' object has no attribute 'add_default_arg'
/ROWAN/group/prod/LINUX_WORKSTATION/acs/spack/spack-0.10.0/lib/spack/spack/build_environment.py:567, in child_execution:
     547      def child_execution(child_connection, input_stream):
     548          try:
     549              setup_package(pkg, dirty=dirty)
     550              function(input_stream)
     551              child_connection.send(None)
     552          except StopIteration as e:
     553              # StopIteration is used to stop installations
     554              # before the final stage, mainly for debug purposes
     555              tty.msg(e.message)
    556              child_connection.send(None)
     557          except:
     558              # catch ANYTHING that goes wrong in the child process
     559              exc_type, exc, tb = sys.exc_info()
     560  
     561              # Need to unwind the traceback in the child because traceback
     562              # objects can't be sent to the parent.
     563              tb_string = traceback.format_exc()
     564  
     565              # build up some context from the offending package so we can
     566              # show that, too.
  >> 567              package_context = get_package_context(tb)
     568  
     569              build_log = None
     570              if hasattr(pkg, 'log_path'):
     571                  build_log = pkg.log_path
     572  
     573              # make a pickleable exception to send to parent.
     574              msg = "%s: %s" % (str(exc_type.__name__), str(exc))
     575  
     576              ce = ChildError(msg, tb_string, build_log, package_context)
     577              child_connection.send(ce)
     578  
     579          finally:
     580              child_connection.close()


Traceback (most recent call last):
  File "/ROWAN/group/prod/LINUX_WORKSTATION/acs/spack/spack-0.10.0/lib/spack/spack/build_environment.py", line 549, in child_execution
    setup_package(pkg, dirty=dirty)
  File "/ROWAN/group/prod/LINUX_WORKSTATION/acs/spack/spack-0.10.0/lib/spack/spack/build_environment.py", line 492, in setup_package
    set_compiler_environment_variables(pkg, spack_env)
  File "/ROWAN/group/prod/LINUX_WORKSTATION/acs/spack/spack-0.10.0/lib/spack/spack/build_environment.py", line 227, in set_compiler_environment_variables
    load_module(mod)
  File "/ROWAN/group/prod/LINUX_WORKSTATION/acs/spack/spack-0.10.0/lib/spack/spack/build_environment.py", line 129, in load_module
    modulecmd.add_default_arg('python')
AttributeError: 'NoneType' object has no attribute 'add_default_arg'
```

### PR DESCRIPTION
**Pull request by becker33**
Fixes #2924 

Supersedes #3111 (includes it with additional features)

Thanks to @skosukhin for #2426, which influenced this PR. 

Changes:
- The function to parse module lines for rpaths now can parse modules in both TCL and lua
- The module manipulation code has been refactored into spack.util
- We can derive the modulecmd executable either from the name 'modulecmd' or from the bash definition of the 'module' function
- The modulecmd executable can have required default args before python (see comment by @Exteris on #2924 for use case)

### CODE REVIEW

**Comment by tgamblin** [comment]
@becker33: this needs tests.

**Comment by becker33** [comment]
Some of the pre-existing methods like `load_module` and `get_path_from_module` we won't be able to test on a travis instance that doesn't have any module program installed (unless we make installing lmod part of the test suite and then use lmod). I've been working on tests for the main functionality (getting the module function from bash) and I think I have that reasonably well tested now.

Travis seems not to be kicking off though, and I unsuccessfully tried to make it start by closing and reopening.

**Comment by luigi-calori** [comment]
I was willing to test this PR but unfortunately it currently has conflict, so I' m unable to test.
Would be possible to rebase while waiting for  merge ?

**Comment by becker33** [comment]
@luigi-calori I've rebased this on current develop, feel free to test it out and let me know what you think!

File: lib/spack/spack/util/module_cmd.py [hunk 1]
```diff
@@ -41,6 +41,8 @@ def get_module_cmd():
 
 def get_module_cmd_from_which():
     module_cmd = which('modulecmd')
+    if not module_cmd:
+        raise ModuleError('`which` did not find any modulecmd executable.')
```
Comment (adamjstewart) [inline lib/spack/spack/util/module_cmd.py]: Might as well use `required=True` in the `which`command. It does all this for you.

Comment (becker33) [inline lib/spack/spack/util/module_cmd.py]: Learn something new every day!

I don't love that error message for this spot, but I agree it's better. I'll change that.

**Comment by becker33** [comment]
@tgamblin After fighting with Travis for too long, I think this is ready now.

**Pull request merged**